models/issue_assignees.go

```diff
@@ -110,73 +110,63 @@ func clearAssigneeByUserID(sess *xorm.Session, userID int64) (err error) {
 	return
 }
 
-// AddAssigneeIfNotAssigned adds an assignee only if he isn't aleady assigned to the issue
-func AddAssigneeIfNotAssigned(issue *Issue, doer *User, assigneeID int64) (err error) {
-	// Check if the user is already assigned
-	isAssigned, err := IsUserAssignedToIssue(issue, &User{ID: assigneeID})
-	if err != nil {
-		return err
-	}
-
-	if !isAssigned {
-		return issue.ChangeAssignee(doer, assigneeID)
-	}
-	return nil
-}
-
 // UpdateAssignee deletes or adds an assignee to an issue
 func UpdateAssignee(issue *Issue, doer *User, assigneeID int64) (err error) {
-	return issue.ChangeAssignee(doer, assigneeID)
+	_, _, err = issue.ChangeAssignee(doer, assigneeID)
+	return err
 }
 
 // ChangeAssignee changes the Assignee of this issue.
-func (issue *Issue) ChangeAssignee(doer *User, assigneeID int64) (err error) {
+func (issue *Issue) ChangeAssignee(doer *User, assigneeID int64) (removed bool, comment *Comment, err error) {
 	sess := x.NewSession()
```

Comment:
This is an opinion of mine, and I can be convinced otherwise. I'm not sure about moving the assignee creation outside of the _transaction_ for issue creation . If we do it this way (which has some good points as well), and there is an error after committing the issue, there will be an 500 Server Error page returned to the user, but the issue will have been created. To address this I'd see to changing the way the page works by returning the created issue and an error flash instead of the 500 Server Error page (e.g. "unable to add assignees"). That would prevent the user from creating the issue a second time. Conceptually, it looks like material for another PR, but this PR is already changing the transaction order, so...

Otherwise we could be using the new `models.DBContext` type and move all this up the chaing, but the issue insert retry code must be moved up the chain as well.

Reply:
I need the "assigned comment" of each added assignee when doing the notification. So if moving this down to models again I need to move the notification part as well.

Also see my [comment below](https://github.com/go-gitea/gitea/pull/8546#issuecomment-543864362) regarding notification and order. 

DBContext means a transaction you do in whole or it will be rewinded?

Reply:
A note is that issue events (like added assignees) does not have an id in the html ("issuecomment-X") so the href hash part doesn't work. But that could easily be added.

Reply:
I'm sorry, I've missed these comments. Sometimes I come in from "Notifications" and sometimes from my e-mail backlog... with different degrees of success. 😝 

I understand your problem. I have no objections about taking the add assignee part out of the issue.